### ISSUE TITLE
CI for azure-functions-maven

### ISSUE THREAD

**Issue by pragnagopa** [Microsoft/azure-maven-plugins#244]
Set up CI that runs E2E tests and verifies plugin  updates for every release.

**Comment by pragnagopa** [comment]
cc: @brunoborges  @JonathanGiles
@jdneo - This came up in our sync up today.  What is the testing process for maven plugin releases?

**Comment by jdneo** [comment]
We have a basic integration test case here: https://github.com/Microsoft/azure-maven-plugins/tree/develop/azure-functions-maven-plugin/src/it

Before each time we release the plugin. Our testers will also have a manual tests according to our test plan here: https://github.com/Microsoft/azure-maven-plugins/wiki/Test-Plan-for-Function

**Comment by jdneo** [comment]
Closing this issue since it's inactive for a long time.

**Comment by pragnagopa** [comment]
cc: @paulbatum 
@jdneo - We do need automated CI for GA. Here is the information I shared in the email

Here some information on how to get started with automation:
- Set environment variables
   - AzureWebJobsScriptRoot to the folder with functions and host.json
  - [samples wiki](https://github.com/Azure/azure-functions-host/wiki#samples) has full list of environment variables for each azure resource.
  - Copy functions runtime release build, here is an example of how the build folder looks like:
	https://ci.appveyor.com/api/buildjobs/tph73ogbmwa39vc4/artifacts/Functions.Private.2.0.11912-alpha.zip
  - Start functions runtime via dotnet cli: 	Run "C:\Program Files (x86)\dotnet\dotnet.exe" Microsoft.Azure.WebJobs.Script.WebHost.dll  this will start the functions runtime
  - You can use [azure java storage sdk](https://github.com/Azure/azure-storage-java) to automate 
storage tests to:
    - Creates a queue
    - Inserts into a queue
    - Verifies function is triggered – You can verify this by adding an output binding that writes to a another queue and reading the second queue

Please feel free to create separate smaller work items for automation and let me know if you need any information from me.

Please note that automation is critical part of release pipeline. For every functions runtime release, I will send a PR updating the runtime version on the build script used by the CI. This will give us coverage for release testing.

**Comment by jdneo** [comment]
OK, we already have some tasks in the current [iteration plan](https://github.com/Microsoft/azure-maven-plugins/issues/349). Let's do this in the next iteration plan.

**Comment by pragnagopa** [comment]
Any udpate on this? We are tracking this for GA. It takes some time to stabilize test automation. Can you please start looking into this? 
Let me know if you run into any issues or have any questions

**Comment by jdneo** [comment]
@Flanker32 Please take a look at this issue.

**Comment by Flanker32** [comment]
@pragnagopa could you please share me more infos about running e2e test? I meet some problems when i try to do it manually.
* Should we set some env about function runtime?  
* After all the  environment variables are set,  I download the runtime build and open the folder ，execute `dotnet.exe Microsoft.Azure.WebJobs.Script.WebHost.dll`, it shows `Unhandled Exception: System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (Exception from HRESULT: 0x8007000B)`. It seems that the version you shared is a X86 build, could you please share me the x64 one, or i should do something to avoid this problem?
* According to [example scripts directory](https://github.com/Azure/azure-functions-host/tree/master/sample/Java) the wiki, functions folders structure should like
```
- functionName
      |- function.java
      |- functionName-Version.jar
      |- function.json
- host.json
```
when i execute mvn package on the test project generated from archetype, the target folder structure are
```
- functionName
      |- function.json
- host.json
- artifactId-Version.jar
- local.settings.json
```
so is there anything we should do after mvn package, is there any documents about it?

**Comment by Flanker32** [comment]
@pragnagopa Here is the plan for the e2e test
* Init function host
  * Clone [Function host](https://github.com/Azure/azure-functions-host) ,build it through powershell script(build.ps1)
  * Unzip `buildoutput\Functions.Private.*.*.*.no-runtime.zip` as function host
  * Clone [Function java worker](https://github.com/Azure/azure-functions-java-worker), build it, and replace the java worker of function host with the output of java worker
  * Set necessary environment value `FUNCTIONS_WORKER_RUNTIME = "java"` `AZURE_FUNCTIONS_ENVIRONMENT = "development"`
* Create function project
  * Create function project by azure-maven-archetypes 
  * *Add more function through maven plugin* 
  * `mvm clean package`
* Run function host and verify whether the function could work normally
  * Set `AzureWebJobsScriptRoot = {path of function project}\target\azure-functions\functionname}`
  * Start function host through `dotnet.exe {path of function host}\Microsoft.Azure.WebJobs.Script.WebHost.dll`
  * Wait for the function start
  * Verify whether function works well through groovy or other script
  * Stop function host 

Feel free to share your thoughts about the test plan 😄

**Comment by pragnagopa** [comment]
- For Functions Host, please see [azure-functions-java-worker appveyor.yml#L59](https://github.com/Azure/azure-functions-java-worker/blob/dev/appveyor.yml#L59), You only need to test with the latest production bits, No need to clone and build. For every runtime release I will send a PR updating the location of the runtime zip.

- For Java worker, same do not clone and build.  Java worker comes as part of the Functions Host. CI for azure-functions-java-worker will take care of testing the worker with the latest mvn plugin

- You should focus on
  - Testing maven archetype i.e build and install maven archetype locally as part of the build script and then create function project
  - Testing maven plugin, same here build and install maven plugin 

- You have the right steps for starting host and running tests. Please see https://github.com/Azure/azure-functions-java-worker/blob/dev/appveyor.yml#L76 for reference on runnning tests.

**Comment by soninaren** [comment]
@Flanker32, As @pragnagopa mentioned.

> @pragnagopa Here is the plan for the e2e test
> 
> * Init function host
>   
>   * Clone [Function host](https://github.com/Azure/azure-functions-host) ,build it through powershell script(build.ps1)
>   * Unzip `buildoutput\Functions.Private.*.*.*.no-runtime.zip` as function host
>   * Clone [Function java worker](https://github.com/Azure/azure-functions-java-worker), build it, and replace the java worker of function host with the output of java worker
>   * Set necessary environment value `FUNCTIONS_WORKER_RUNTIME = "java"` `AZURE_FUNCTIONS_ENVIRONMENT = "development"`

**No need to do this step. We automatically upload the CLI from every build to Azure Storage. You can download the latest CLI from** https://functionsclibuilds.blob.core.windows.net/builds/2/latest/Azure.Functions.Cli.{platform}.zip. 

**Replace {platform} with one of these values linux-x64, osx-x64, win-x64, win-x86 based on what configuration you are using. Also you can get the corresponding CLI version number from here**  https://functionsclibuilds.blob.core.windows.net/builds/2/latest/version.txt

> * Create function project
>   
>   * Create function project by azure-maven-archetypes
>   * _Add more function through maven plugin_
>   * `mvm clean package`
> * Run function host and verify whether the function could work normally
>   
>   * Set `AzureWebJobsScriptRoot = {path of function project}\target\azure-functions\functionname}`
>   * Start function host through `dotnet.exe {path of function host}\Microsoft.Azure.WebJobs.Script.WebHost.dll`
>   * Wait for the function start
>   * Verify whether function works well through groovy or other script
>   * Stop function host
> 
> Feel free to share your thoughts about the test plan 😄

**Comment by pragnagopa** [comment]
Thanks @soninaren. @Flanker32 - Please see https://github.com/Azure/azure-functions-java-worker/blob/dev/appveyor.yml#L49 for reference on how to download and build E2E tests using the CLI and run tests.

**Comment by Flanker32** [comment]
@pragnagopa  @soninaren 
Thanks for your efforts. I tried to download cli from https://functionsclibuilds.blob.core.windows.net/builds/2/latest/Azure.Functions.Cli.{platform}.zip, but when i execute `dotnet.exe {path of function host}\Microsoft.Azure.WebJobs.Script.WebHost.dll`, it shows 
``` powershell
Unhandled Exception: System.IO.FileNotFoundException: Could not find file 'C:\Users\hanli\Downloads\Azure.Functions.Cli.win-x64\Microsoft.Azure.WebJobs.Script.WebHost.deps.json'.
```
I can't find the json as well as runtime files in this build, so i clone the function host repo and build it, that could work. Or is there any solution to solve this problem?

**Comment by soninaren** [comment]
@Flanker32 you are running the wrong command.
This is what you need to run
```
{CLI download path}\func.exe host start
```

Here is the powershell command for it
https://github.com/Azure/azure-functions-java-worker/blob/dev/appveyor.yml#L85

**Comment by jdneo** [comment]
Implemented in #425
